Door and teleport improvements

Brockengespenst opened the pull request:
Provides the following changes to door and teleport transitions:
- ShrinkFade is now used as fade effect instead of FadeToBlack
- Player controls are disabled while fading out to prevent Tux from walking away the door while it opens
- Tux walks automatically in front of door if up key is pressed and Tux is not directly in front of the door
- Tux cannot be hurt by enemies while controls are disabled
- Added some time of invincibility when spawning at the target door
- Added scripting support for Level.spawn_transition

Default fade effect when entering a GameSession is now the ShrinkFade effect.

Fixes #2409
Fixes #2649

Brockengespenst commented:
Right now the same approach as the safe timer after getting hurt is used when doing a teleport using doors (except for Tux doing the "blinking"). The same amount of time (1.8 sec) is used - which might be a bit long.
The scripting part is preliminary, just for testing. It adds a command `Level.spawn_transition("[sector]", "[spawnpoint]", "[transition_type]");` with transition type  `"circle"` (for shrink fade), `"fade"` (fade to black) or any other string will be interpreted as none, which means no fade effect at all.

Brockengespenst commented:
For local multiplayer a proper strategy might be needed to determine the focus for the shrink fade when the script is used. Right now the first alive player will be chosen. This should work for single player, but for local multiplayer another approach might be better. Suggestions are welcome.

tobbi commented on src/supertux/game_session.hpp:
```diff
@@ -88,6 +95,11 @@ class GameSession final : public Screen,
+  void respawn_with_fade(const std::string& sectorname,
+                         const std::string& spawnpointname,
+                         const FadeType fade_type,
+                         const Vector fade_point,
```
I wonder if passing by reference here is preferable.

Vankata453 replied:
I think so.

Brockengespenst replied:
Good idea.

tobbi commented on src/supertux/game_session.hpp:
```diff
@@ -88,6 +95,11 @@ class GameSession final : public Screen,
+                         const std::string& spawnpointname,
+                         const FadeType fade_type,
+                         const Vector fade_point,
+                         const bool make_invincible = false);
```
Does `const bool` mean that make_invincible is always false?

Vankata453 replied:
I don't think those two relate. It's just that the parameter can't be modified.

Brockengespenst replied:
Indeed, it just says that the parameter cannot be altered inside the function body.

tobbi commented:
Sorry, I thought this was already ready for review, thus adding my comments. You can ignore them if you like. No pressure.

Vankata453 commented on src/supertux/game_session.hpp:
```diff
@@ -76,6 +76,13 @@ class GameSession final : public Screen,
   };
 
 public:
+  enum class FadeType
+  {
+    NONE,
+    FADE,
+    CIRCLE
+  };
```
Maybe these enumerators would be better suited in the `ScreenFade` class ("src/supertux/screen_fade.hpp)?

Brockengespenst replied:
Yes, I will move it there.

Vankata453 commented on src/supertux/game_session.cpp:
```diff
@@ -475,6 +515,35 @@ GameSession::update(float dt_sec, const Controller& controller)
+      ScreenManager::current()->set_screen_fade(std::make_unique<ShrinkFade>(shrinkpos, FADE_TIME, ShrinkFade::FADEIN));
+      break;
+    }
+    case FadeType::NONE:
+    default:
```
`default:` would take care of `NONE` either way.

Vankata453 commented on src/supertux/game_session.cpp:
```diff
@@ -475,6 +515,35 @@ GameSession::update(float dt_sec, const Controller& controller)
 
+    switch (m_spawn_fade_type)
+    {
+    case FadeType::FADE:
+    {
+      ScreenManager::current()->set_screen_fade(std::make_unique<FadeToBlack>(FadeToBlack::FADEIN, FADE_TIME));
+      break;
+    }
+    case FadeType::CIRCLE:
+    {
+      const Vector spawn_point_position = sector->get_spawn_point_position(m_newspawnpoint);
+      const Vector shrinkpos = spawn_point_position - sector->get_camera().get_translation();
+
+      ScreenManager::current()->set_screen_fade(std::make_unique<ShrinkFade>(shrinkpos, FADE_TIME, ShrinkFade::FADEIN));
+      break;
+    }
+    case FadeType::NONE:
+    default:
+      break;
```
Indent the `switch` statement cases two more spaces in.

Vankata453 commented on src/supertux/game_session.cpp:
```diff
@@ -282,6 +298,23 @@ GameSession::on_escape_press(bool force_quick_respawn)
   }
 }
 
+Vector
+GameSession::get_fade_point() const
+{
+  // Get first player that is alive
+  // Should work for single player without problems,
+  // but for multiplayer a proper handling needs to be done
+  for (const auto& player : m_currentsector->get_players())
+  {
+    if (!player->is_dead() && !player->is_dying())
+    {
+      return player->get_bbox().get_middle();
+    }
+  }
+
+  return Vector(0.0f, 0.f);
+}
+
```
Perhaps the fade could be done where the door is, instead of where the first alive player is?

For the script, a parameter as to where to do the fade at could work.

Brockengespenst replied:
This function is only used when the point cannot be determined automatically. If there is a door, that triggers the respawn, its coordinates are used (see door passing its center to the respawn_with_fade function).
This functions is mainly for the scripting part where the trigger cannot be determined (or at least I did not figure out how to get it in case of the "Level.*" commands or if invoked from console.

Vankata453 replied:
For the scripting function, screen-relative coordinates could be required.

Rusty-Box commented:
This is lovely! Though I do think Tux should not be able to receive any further inputs during the teleport transition + door transition (kinda like when using the scrippted controller script). Imo. it does feel kinda weird being able to keep on going during it.

MatusGuy commented on src/supertux/shrinkfade.cpp:
```diff
@@ -41,7 +42,7 @@ void
-  float diameter = 2 * m_initial_size * (1.0f - progress);
+  float diameter = 2 * m_initial_size * (m_direction == FADEOUT ? (1.0f - progress) : progress);
   context.color().draw_inverse_ellipse(m_dest, Vector(1.1f * diameter, diameter),
                                          Color(0, 0, 0), LAYER_GUI + 1);
```
Don't forget to change the layer to something lower than LAYER_GUI or else the effect goes above the pause menu gui. I recommend LAYER_GUI-5

Brockengespenst replied:
Thanks for the hint!

Rusty-Box commented:
Can we also have teleportation transition make you immune to death, like falling offscreen? For example if you want Tux to jump in a hole in the ground at the bottom of the screen, teleport, without having him die.

Brockengespenst commented:
> Though I do think Tux should not be able to receive any further inputs during the teleport transition + door transition (kinda like when using the scrippted controller script). Imo. it does feel kinda weird being able to keep on going during it.

I totally agree. First, I was playing around with setting the gamespeed to zero, but than then fade animation was stopped as well. I can check if blocking the controls works better. I think in that case it would also make sense to make Tux invincible already at the point when the controls are blocked, before the actual teleport is done.
Btw. right now Tux is invincible only when going through a door. Should this be applied to Level.spawn_transition scripting as well?

> Can we also have teleportation transition make you immune to death, like falling offscreen? For example if you want Tux to jump in a hole in the ground at the bottom of the screen, teleport, without having him die.

You mean before the actual teleport is done? I was not aware that this might be a use case, but I will check.

Rusty-Box commented:
> Btw. right now Tux is invincible only when going through a door. Should this be applied to Level.spawn_transition scripting as well?

I think so yeah. Also, the time Tux remains invincible after the door is indeed too long. maybe shorten it to like 1 is probably more appropriate.

> You mean before the actual teleport is done? I was not aware that this might be a use case, but I will check.

To be fair, it happens rather rarely to be used in this way but can be quite useful, I'd say. Basically once the script thus the transition is triggered Tux should not be affected by "death barriers" etc.

Brockengespenst commented:
I shortened the safe time to 1 sec now and disabled the controls for the player that uses the door (if in mulitplayer, the other players are not affected) until the fadout part of the effect is done. After actual teleport they are reactivated. Tux is also safe while controls are disabled.

Regarding the "falling off screen": When calling scripts like Level.spawn*, the information which player triggered the script is lost quite early in the chain, so it is not that easy to find out which player this should affect. If making all players invincible, I think it might lead to unwanted behavior in case of multiplayer.

I am wondering if it would make sense to introduce an own trigger type (similar to the door, but without being visible and all the animation stuff), because this would solve some problems: The player that should be invincible could be determined, as well as the point for the fade effect.
It would also be easier to make this configurable, e.g. 
* Target sector / spawnpoint
* Chosing the transition type from a selection menu
* Either checkbox if invincibility after respawn should be applied or provide option to specify the time of invincibility 
* Checkbox if Tux should be invincible until actual teleport (to allow him to fall off from screen while fade out effect is in progress)

What do you think?

Rusty-Box commented:
While the transition effect no longer blocks the menu text it still blocks the textbox of it.

Also, one question I have: is it possible to make the circle of the "circle transition" in the center of the screen when the Camera mode is set to manual? I noticed that during cutscenes where the player object is often offscreen, it looks weird having it anchored to the player. All these you a "manual" camera.

Rusty-Box commented:
> I am wondering if it would make sense to introduce an own trigger type (similar to the door, but without being visible and all the animation stuff), because this would solve some problems: The player that should be invincible could be determined, as well as the point for the fade effect. It would also be easier to make this configurable, e.g.

Honestly, I think I wouldn't mind that.

Brockengespenst commented:
> While the transition effect no longer blocks the menu text it still blocks the textbox of it.

Thanks, should be fixed now.

> Also, one question I have: is it possible to make the circle of the "circle transition" in the center of the screen when the Camera mode is set to manual? I noticed that during cutscenes where the player object is often offscreen, it looks weird having it anchored to the player. All these you a "manual" camera.

While in cutscenes the center of screen is now chosen. Looked indeed a bit weird.
For the multiplayer I changed the fade point to the middle of all alive players, this feels a bit more fair. But I am open to better ideas.

Open point would be the invinvibility for "falling" off screen. So when you say you do not mind to introduce a separate trigger to determine the player that actually started the script, shall all players be invincible in case of multiplayer?

Brockengespenst commented:
I agree that it would make sense to only make the player that triggered the script invincible, but that would require too much rework in the scripting part. Maybe this will become easier once the SimpleSquirrel PR from @Vankata453 is merged.
So I added it for all players right now, so you can fall off from screen right now without being killed when Level.spawn_transition is invoked.

I will have a look at the Squirrel Enumeration for the fade type parameter now.

tobbi commented:
Hey, @Brockengespenst We were wondering whether you'd be willing to join our SuperTux server (external/internal) one so we can coordinate / communicate better?

tobbi commented:
> You should mention that it is a discord server specifically

You're right. I changed it. Thought I'd written that.

Rusty-Box commented:
So it appears that the center transition doesn't seem to work quite with levels using the `Camera.set_sclae(x);` script. It look quite offset. Also, I still think the centered effect should apply when Camera Mode is "Manual".

This is a cutscene that uses a zoomed in Camera. As you you can see the effect is not fully centered:

![](https://github.com/SuperTux/supertux/assets/20615048/d950299a-9de5-4a12-a80f-981f8d0d70ef)

---

![](https://github.com/SuperTux/supertux/assets/20615048/e74f3356-695a-429d-abbf-a67740973516)

This is one of the text scroller "levels". It uses only a manual Camera as I'm not sure how to really use the cutscene scripts here, I was hoping we could have that settled through the manual Camera, having the effect centered. Right now it still does focus Tux who's meant to be hidden in this scene.

Brockengespenst commented:
I will have a look a the scaling and the manual camera (seems I mixed this up with cut scenes, sorry for that).

Rusty-Box commented:
> [...] (seems I mixed this up with cut scenes, sorry for that).

It's okay. They usually go hand in hand. But not for those text scrollers. :D

Brockengespenst commented:
Camera scale should now be taken into account. Also for camera in manual mode, the middle of the screen should be used now.

Rusty-Box commented:
Very nice! One last thing regarding the door. I find it a bit annoying that I have to wait until the door fully opens before Tux "freezes" and the transition begins to play. In my opinion, once you press up, Tux should freeze and the transition play while the door is opening

Brockengespenst commented:
> In my opinion, once you press up, Tux should freeze and the transition play while the door is opening

Makes sense, I will change the it.

Rusty-Box commented:
Okay I said "last one" but, uh, the coin, enemy and secret counter during the pause screen, the timer, coin and powerup counter, as well as "Show Controller, Show Framerate, Show Player Position" overlay/text get covered by the transition effect. Sorry about that 😅

Brockengespenst commented:
I adjusted the layers so that the fade effect will not cover the player hud anymore. I noticed that the InfoBlock used a GUI layer which caused it not to be covered by the fade effect (see rooted forest, first level, just the one before the door). I would consider InfoBlocks as part of the game and not of the GUI, so I moved it into a lower layer. It seems that it does not have any side effects, but please check.

Brockengespenst commented:
I would have another small feature for this: If Tux is not standing close to the center of the door when up is pressed, I could make him automatically move closer to the door using the scripted controller. Code would be ready to push, if you appreciate this feature.

Rusty-Box commented:
> I would have another small feature for this: If Tux is not standing close to the center of the door when up is pressed, I could make him automatically move closer to the door using the scripted controller. Code would be ready to push, if you appreciate this feature.

Depending on how it turns out I would not be against this, honestly.

Brockengespenst commented:
Give it a try. If you do not like it, I will revert it.

Rusty-Box commented:
I really like that! This is cool. For once Tux looks like he actually is going through that door :D


> I am wondering if it would make sense to introduce an own trigger type (similar to the door, but without being visible and all the animation stuff), because this would solve some problems: The player that should be invincible could be determined, as well as the point for the fade effect. It would also be easier to make this configurable, e.g.


I assume, if this would be something that gets "greenlit" this would not be part of this PR? If not I would go ahead and approve this PR

Brockengespenst commented:
> I assume, if this would be something that gets "greenlit" this would not be part of this PR?

It could be done independently of this PR, if someone needs this feature.

Right now the last scripting parameter for Level.spawn_transition (determining the fade effect type) is still a string and not yet an enumeration. To be honest, I have not had a look how to teach Squirrel a new Enumeration, yet.

Rusty-Box commented:
> It could be done independently of this PR, if someone needs this feature.

Either way I probably should. That way we can merge this sooner rather than later. So yeah, best to just make that its own PR.

> Right now the last scripting parameter for Level.spawn_transition (determining the fade effect type) is still a string and not yet an enumeration. To be honest, I have not had a look how to teach Squirrel a new Enumeration, yet.

Does it need to be though? I think there's a couple of scripts that deal with strings for certain things, like the easing on the `Camera.ease_scale(x, t, "EaseMode");` script. Personally I think the way it is now is totally fine, but perhaps I am missing something?

Brockengespenst commented:
@Vankata453 suggested this. If he is fine with the current solution, I would be fine, too.

Vankata453 commented:
> Does it need to be though?

Looking at it, an enumerator is not necessarily needed, the possible strings would be documented either way.

Rusty-Box commented:
I found one major bug. Since you use scripted controller, Tux drops whatever he is carrying while going into a door!

Brockengespenst commented:
Uh, thanks for the hint! Sorry, I will have a look.

Brockengespenst commented:
I removed the usage of the scripting controller and used other mechanism instead. Tested with lantern and it did not drop anymore when going through the door.

Vankata453 commented on src/supertux/game_session.cpp:
```diff
@@ -282,6 +298,58 @@ GameSession::on_escape_press(bool force_quick_respawn)
+    }
+  }
+
+  fade_point = (fade_point - m_currentsector->get_camera().get_translation()) * m_currentsector->get_camera().get_current_scale();
+
+  return fade_point;
```
The result could be `return`ed directly.

Vankata453 commented on src/supertux/game_session.hpp:
```diff
@@ -34,6 +34,7 @@
 #include "supertux/sequence.hpp"
 #include "supertux/timer.hpp"
 #include "video/surface_ptr.hpp"
+#include "supertux/screen_fade.hpp"
```
Should go upwards for alphabetical sorting.

Vankata453 commented on src/supertux/game_session.hpp:
```diff
@@ -135,12 +141,17 @@ class GameSession final : public Screen,
 
   bool m_prevent_death; /**< true if players should enter ghost mode instead of dying */
 
+  static constexpr float TELEPORT_FADE_TIME = 1.0f; /**< Duration of teleport fade animation */
+
```
Instead, use a static variable, located at the start of the class declaration, prefixed with `s_`:

```
public:
  static const float s_teleport_fade_time;
```

The static variable should be defined at the top of the .cpp file:

```
const float GameSession::s_teleport_fade_time = 1.f;
```

Brockengespenst replied:
Sorry, somehow I missed this. It could be moved completely to the implementation file as the GameSession will make Tux safe during the fadeout animation anyway. I hope it is Ok that I did it similar to the other static const's in the other code instead of using your suggestion. Otherwise I can change it again.

Vankata453 replied:
It's okay, my suggestion was keeping in mind usage from other files.

Vankata453 commented on src/supertux/sector.hpp:
```diff
@@ -47,6 +47,7 @@ class Size;
 class TextObject;
 class TileMap;
 class Writer;
+class SpawnPointMarker;
```
Should be moved upwards (alphabetical).

tobbi commented on src/supertux/game_session.cpp:
```diff
@@ -282,6 +298,58 @@ GameSession::on_escape_press(bool force_quick_respawn)
+    }
+  }
+
+  fade_point = (fade_point - m_currentsector->get_camera().get_translation()) * m_currentsector->get_camera().get_current_scale();
```
Maybe move camera into a variable before this?

Vankata453 commented on src/supertux/shrinkfade.cpp:
```diff
@@ -41,9 +42,9 @@ void
+  float diameter = 2 * m_initial_size * (m_direction == FADEOUT ? (1.0f - progress) : progress);
   context.color().draw_inverse_ellipse(m_dest, Vector(1.1f * diameter, diameter),
-                                         Color(0, 0, 0), LAYER_GUI + 1);
+                                         Color(0, 0, 0), LAYER_LIGHTMAP - 1);
```
These transitions are not only used in levels, but also in the GUI as well. I wonder if we could use a variable here to know which is the case?

tobbi commented on src/supertux/sector.cpp:
```diff
@@ -179,6 +179,29 @@ Sector::activate(const std::string& spawnpoint)
+  return sp;
+}
+
+
```
Extraneous line. (I'm nitpicking here, I know)

Vankata453 commented on src/supertux/shrinkfade.hpp:
```diff
@@ -24,7 +24,8 @@
 {
 public:
-  ShrinkFade(const Vector& point, float fade_time);
+  enum Direction { FADEOUT, FADEIN };
+  ShrinkFade(const Vector& point, float fade_time, Direction = FADEOUT);
```
Separating enumerators from constructor:

```suggestion
  enum Direction { FADEOUT, FADEIN };

public:
  ShrinkFade(const Vector& point, float fade_time, Direction = FADEOUT);
```

tobbi commented:
Looks good otherwise. Just those few changes.

Vankata453 commented on src/trigger/door.cpp:
```diff
@@ -29,34 +29,43 @@
 #include "supertux/flip_level_transformer.hpp"
 #include "util/reader_mapping.hpp"
 
+
+static const float STAY_OPEN_TIME = 1.0f;
+static const float LOCK_WARN_TIME = 0.5f;
+static const float CENTER_EPSILON = 5.0f;
+static const float WALK_SPEED = 100.0f;
+
+
```
You could remove the excess newlines, if you want to.

Vankata453 commented on src/scripting/level.cpp:
```diff
@@ -45,6 +45,21 @@ Level_spawn(const std::string& sector, const std::string& spawnpoint)
+
+  ScreenFade::FadeType fade_type = ScreenFade::FadeType::NONE;
+
+  if (transition == "fade")
+    fade_type = ScreenFade::FadeType::FADE;
+  else if (transition == "circle")
+    fade_type = ScreenFade::FadeType::CIRCLE;
```
An additional `else`, which prints out a warning (using `log_warning`) that the transition provided is not valid, would be great and would allow for consistency with other similar methods (for example, ones that require easing).

```suggestion
  if (transition == "fade")
    fade_type = ScreenFade::FadeType::FADE;
  else if (transition == "circle")
    fade_type = ScreenFade::FadeType::CIRCLE;
  else
    log_warning << "Invalid transition type '" << transition << "'." << std::endl;
```

Brockengespenst commented on src/object/player.cpp:
```diff
@@ -2374,6 +2375,13 @@ Player::make_invincible()
+Player::make_temporarily_safe(float safe_time)
+{
+  m_safe_timer.start(safe_time);
+  m_is_intentionally_safe = true;
```
Otherwise Tux would start to blink once he wants to enter the door. This flag is used to avoid the blinking effect.

Brockengespenst commented:
> Makes sense as Tux walks towards the center of the door anyway.

I just noticed that I use the door's center position already for the fade effect.

Rusty-Box commented:
Have requested changes been adressed or... what the status?

Vankata453 commented:
@Rusty-Box I have to re-review and test - will try to soon.

jonathanpoatato commented:
w

Vankata453 commented on src/supertux/shrinkfade.hpp:
```diff
@@ -24,18 +24,23 @@
+  enum Direction { FADEOUT, FADEIN };
+
+public:
+  ShrinkFade(const Vector& point, float fade_time, int draw_layer, Direction = FADEOUT);
```
I suggest having the previous value, used in `draw()`, which is `LAYER_GUI + 1`, as a default.

Brockengespenst replied:
Done.

Vankata453 commented on src/scripting/level.cpp:
```diff
@@ -22,6 +22,8 @@
 #include "supertux/game_session.hpp"
 #include "supertux/sector.hpp"
 
+#include "util/log.hpp"
```
There shouldn't be a separation of includes here:

```suggestion
#include "util/log.hpp"
```

Brockengespenst replied:
Done.

Vankata453 commented on src/scripting/level.hpp:
```diff
@@ -66,6 +66,19 @@ bool Level_has_active_sequence();
+   If that spawnpoint doesn't exist either, Tux will simply end up at the origin (top-left 0, 0).
+ * @param string $sector
+ * @param string $spawnpoint
+ * @param string $transition
```
Valid transition types should be specified under the description of this parameter.

Brockengespenst replied:
Done.

Vankata453 reviewed (commented):
Tested this and it seems to work well!

However, I noticed a few other things on the code side, so I have commented my suggestion for those. There is also one comment from my previous review left unresolved.

Brockengespenst commented:
>     * Currently, entering a `GameSession` always plays the `ShrinkFade` transition. It seems to be intended, so I wonder why it was added.

Sorry, I do not really get the point?

Vankata453 commented:
> Sorry, I do not really get the point?

I was wondering why a circle fade is drawn when entering a level (because I didn't find comments where it has been discussed, may have missed them).

Brockengespenst commented:
> I was wondering why a circle fade is drawn when entering a level (because I didn't find comments where it has been discussed, may have missed them).

From #2409:

> What if instead of simply fading to black (using doors) or no transition at all (using Level.spawn();) we could use transition effect closer to entering a level, the circular one.

I thought that it was intended by @Rusty-Box, that's why I added it.

Vankata453 commented:
> but until now there were no complaints about it

Personally I'm not necessarily against it, but it was a surprise. :D

Vankata453 commented:
In terms of my reviews, one more [comment](https://github.com/SuperTux/supertux/pull/2844#discussion_r1548724420) is left unresolved.

Rusty-Box commented:
> I thought that it was intended by @Rusty-Box, that's why I added it. Maybe I mixed this up with the world map, but until now there were no complaints about it. :-)

It was never mentioned anywhere, yes, but I really like that effect for level start so I'd personally love to keep it around.

MatusGuy reviewed (approved):
Untested.

tobbi commented on src/trigger/door.cpp:
```diff
@@ -85,79 +92,98 @@ Door::after_editor_set()
+    }
+    else
+    {
+      m_triggering_player->walk(0.0);
```
Maybe: `0.0f`? Shouldn't make a difference.

Brockengespenst replied:
Done.

tobbi reviewed (changes requested):
Just the small thing that I marked. Otherwise looking good. Untested, though. I'll test it in a few.

Rusty-Box reviewed (approved):
Apparently I didn't post a proper review here so... yeah, approved!

tobbi commented:
Okay, so a few things that I noticed: Is it necessary to have the teleport transition circle when starting a level? IDK, to me it kinda seems out of place.

Also, I'd like the speed of the transition to pick up a bit. Seems a little bit too slow for me.

Rusty-Box commented:
> Okay, so a few things that I noticed: Is it necessary to have the teleport transition circle when starting a level? IDK, to me it kinda seems out of place.

I think it looks nicer. More cartoony. Fits the game's style more. It also fits since entering a level from the worldmap uses the exact same transition

> Also, I'd like the speed of the transition to pick up a bit. Seems a little bit too slow for me.

Entering a door is quite quicker than it used to be, idk if it needs to be faster. A tad maybe. Personally I think it's fine enough

Vankata453 commented:
> Entering a door is quite quicker than it used to be, idk if it needs to be faster. A tad maybe. Personally I think it's fine enough

I agree, faster than this would be very odd.

tobbi commented:
Okay, in that case: Approval from me.
